### PR TITLE
Deprecate `patterns`

### LINKED ISSUE #1007: Deprecate janitor.patterns
janitor.patterns is not really useful (beauty of hindsight :grin: ) , as it is simply a wrapper around `re.compile`, which a user can import and wrap

```py
from re import compile as patterns # or whatever name user prefers
import janitor
import pandas as pd

df.select_columns(patterns('\d+')
```

### PR DESCRIPTION
**Pull request by samukweku**
# PR Description

Please describe the changes proposed in the pull request:

- Deprecate `patterns` function
- Minor refactor to `select_columns`

**This PR resolves #1007 .**

# PR Checklist

Please ensure that you have done the following:

1. [ ] PR in from a fork off your branch. Do not PR from `<your_username>`:`dev`, but rather from `<your_username>`:`<feature-branch_name>`.

2. [ ] If you're not on the contributors list, add yourself to `AUTHORS.rst`.

3. [ ] Add a line to `CHANGELOG.md` under the latest version header (i.e. the one that is "on deck") describing the contribution.
    - Do use some discretion here; if there are multiple PRs that are related, keep them in a single line.

# Automatic checks

There will be automatic checks run on the PR. These include:

- Building a preview of the docs on Netlify
- Automatically linting the code
- Making sure the code is documented
- Making sure that all tests are passed
- Making sure that code coverage doesn't go down.

# Relevant Reviewers

Please tag maintainers to review.

- @ericmjl

### CODE REVIEW

**Comment by ericmjl** [comment]
🚀 Deployed on https://deploy-preview-1027--pyjanitor.netlify.app

File: midpoint.csv [hunk 1]
```diff
@@ -0,0 +1,3 @@
+,a_new,b
```
Comment (thatlittleboy) [inline midpoint.csv]: Somehow this `midpoint.csv` file always makes its way into the PR 😄  I presume this is a mistake?

File: tests/functions/test_pivot_longer.py [hunk 2]
```diff
@@ -3,7 +3,7 @@
 import pytest
 from pandas.testing import assert_frame_equal
 
-from janitor import patterns
+from re import compile as patterns
```
Comment (thatlittleboy) [inline tests/functions/test_pivot_longer.py]: My opinion is that for internal uses (of `patterns`), e.g. in tests, we don't have to follow any deprecation cycle or the like.
I would suggest to remove all uses of `patterns` and change them to `re.compile` directly.

The code will be clearer for developers to read in the future as well. (They might wonder why we choose to perform this strange alias)

File: tests/utils/test__select_column.py [hunk 3]
```diff
@@ -5,8 +5,8 @@
 import pandas as pd
 import pytest
 
-from janitor import patterns
-from janitor.functions.utils import _select_column_names
+from re import compile as regex
```
Comment (thatlittleboy) [inline tests/utils/test__select_column.py]: Same comment here about being explicit in the use of `re.compile`.

File: tests/utils/test__select_column.py [hunk 4]
```diff
@@ -320,7 +344,7 @@ def test_regex(df1):
     assert _select_column_names(re.compile(r"\d$"), df1) == list(
         df1.filter(regex=r"\d$").columns
     )
-    assert _select_column_names(patterns(r"\d$"), df1) == list(
+    assert _select_column_names(regex(r"\d$"), df1) == list(
         df1.filter(regex=r"\d$").columns
     )
```
Comment (thatlittleboy) [inline tests/utils/test__select_column.py]: Taking the above comment into account, in which case, I suppose this test would be rendered useless (it's essentially a duplicate of the test above, yeah?)

**Review by thatlittleboy — CHANGES_REQUESTED** [review]
All really welcome changes to the code base! Thanks.

I just left 1 comment about the explicit use of `re.compile` within tests. Open to discussion about this.

File: midpoint.csv (hunk 1 above)
Comment (ericmjl) [reply midpoint.csv]: haha yeah, let's have this file removed, @samukweku :smile:

File: tests/functions/test_pivot_longer.py (hunk 2 above)
Comment (ericmjl) [reply tests/functions/test_pivot_longer.py]: Agreed, thanks for raising, @thatlittleboy!

**Review by ericmjl — APPROVED** [review]
Agree with all of @thatlittleboy's comments, but nonetheless, I'm good with pre-approving the PR!

File: tests/utils/test__select_column.py [hunk 5]
```diff
@@ -344,7 +343,7 @@ def test_regex(df1):
     assert _select_column_names(re.compile(r"\d$"), df1) == list(
         df1.filter(regex=r"\d$").columns
     )
-    assert _select_column_names(regex(r"\d$"), df1) == list(
+    assert _select_column_names(re.compile(r"\d$"), df1) == list(
         df1.filter(regex=r"\d$").columns
     )
```
Comment (thatlittleboy) [inline tests/utils/test__select_column.py]: This test looks to be a duplicate of the one above. Can we remove it?

**Review by thatlittleboy — APPROVED** [review]
Approving and merging, thanks!

**Pull request merged**